test/Dialect/Arc/arc-canonicalizer.mlir

```diff
@@ -453,22 +453,14 @@ in %clock: !seq.clock, in %o: i8, in %v: i8, in %q: i8, in %s: i8) {
 }
 
 // CHECK-LABEL: hw.module @Needs_Shuffle(in %b : i8, in %e : i8, in %h : i8, in %k : i8, in %c : i8, in %f : i8, in %i : i8, in %l : i8, in %n : i8, in %p : i8, in %r : i8, in %t : i8, in %en : i1, in %clock : !seq.clock, in %o : i8, in %v : i8, in %q : i8, in %s : i8) {
-// CHECK-NEXT:    [[VEC0:%.+]]:4 = arc.vectorize (%b, %e, %h, %k), (%c, %f, %i, %l) : (i8, i8, i8, i8, i8, i8, i8, i8) -> (i8, i8, i8, i8) {
-// CHECK-NEXT:    ^[[BLOCK:[[:alnum:]]+]](%arg0: i8, %arg1: i8):
-// CHECK-NEXT:      [[OUT:%.+]] = comb.add %arg0, %arg1 : i8
-// CHECK-NEXT:      arc.vectorize.return [[OUT]] : i8
-// CHECK-NEXT:    }
-// CHECK-NEXT:    [[VEC1:%.+]]:4 = arc.vectorize ([[VEC0]]#1, [[VEC0]]#0, [[VEC0]]#2, [[VEC0]]#3), (%n, %p, %r, %t) : (i8, i8, i8, i8, i8, i8, i8, i8) -> (i8, i8, i8, i8) {
-// CHECK-NEXT:    ^[[BLOCK:[[:alnum:]]+]](%arg0: i8, %arg1: i8):
-// CHECK-NEXT:      [[OUT:%.+]] = comb.and %arg0, %arg1 : i8
-// CHECK-NEXT:      arc.vectorize.return [[OUT]] : i8
-// CHECK-NEXT:    }
-// CHECK-NEXT:    [[VEC2:%.+]]:4 = arc.vectorize ([[VEC1]]#1, [[VEC1]]#0, [[VEC1]]#2, [[VEC1]]#3), (%o, %v, %q, %s) : (i8, i8, i8, i8, i8, i8, i8, i8) -> (i8, i8, i8, i8) {
-// CHECK-NEXT:    ^[[BLOCK:[[:alnum:]]+]](%arg0: i8, %arg1: i8):
-// CHECK-NEXT:      [[OUT:%.+]] = arc.call @Just_A_Dummy_Func(%arg0, %arg1) : (i8, i8) -> i8
-// CHECK-NEXT:      arc.vectorize.return [[OUT]] : i8
+// CHECK-NEXT:    [[VEC:%.+]]:4 = arc.vectorize (%b, %e, %h, %k), (%c, %f, %i, %l), (%n, %p, %r, %t), (%o, %v, %q, %s) : (i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8) -> (i8, i8, i8, i8) {
```

Comment:
You need to make sure that all input vectors are re-ordered the same way. As we see here, the vector `(%n, %p, %r, %t)` didn't change, but should be `(%p, %n, %r, %t)` after the optimization.

Reply:
Oops, I didn't notice this, I will think of a way to do this, maybe bind the operands to index and shuffle them?

We will need more optimizations here to reduce the shuffle cost. 
I think this optimization will affect the canonicalization of `keeping one vector if the pattern is repeated.` as it will treat `[a, b, c, d]` differently from `[a, c, b, d]`.